hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,395 @@
+    if (cpHost != null) {
+      cpHost.postSnapshot(ProtobufUtil.createSnapshotDesc(snapshot), htd);
+    }
+  }
+
+  private void verifySnapshot(MasterProcedureEnv env) throws IOException {
+    int verifyThreshold = env.getMasterConfiguration()
+      .getInt("hbase.snapshot.remote.verify.threshold", 10000);
+    int numRegions = (int) env.getAssignmentManager()
+      .getTableRegions(snapshotTable, false)
+      .stream().filter(r -> RegionReplicaUtil.isDefaultReplica(r)).count();
+
+    if (numRegions >= verifyThreshold) {
+      addChildProcedure(createRemoteVerifyProcedures(env));
+    } else {
+      MasterSnapshotVerifier verifier =
+        new MasterSnapshotVerifier(env.getMasterServices(), snapshot, workingDirFS);
+      verifier.verifySnapshot();
+    }
+  }
+
+  private void completeSnapshot(MasterProcedureEnv env) throws IOException {
+    // complete the snapshot, atomically moving from tmp to .snapshot dir.
+    SnapshotDescriptionUtils.completeSnapshot(snapshotDir, workingDir,
+      env.getMasterFileSystem().getFileSystem(), workingDirFS, conf);
+    // update metric. when master restarts, the metric value is wrong
+    metricsSnapshot.addSnapshot(status.getCompletionTimestamp() - status.getStartTime());
+    if (env.getMasterCoprocessorHost() != null) {
+      env.getMasterCoprocessorHost()
+        .postCompletedSnapshotAction(ProtobufUtil.createSnapshotDesc(snapshot), htd);
+    }
+    status.markComplete("Snapshot " + snapshot.getName() + "  completed");
+  }
+
+  private void snapshotOfflineRegions(MasterProcedureEnv env) throws IOException {
+    List<RegionInfo> regions = env.getAssignmentManager()
+      .getTableRegions(snapshotTable, false).stream()
+      .filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .filter(RegionInfo::isSplit).collect(Collectors.toList());
+
+    ThreadPoolExecutor exec = SnapshotManifest
+      .createExecutor(env.getMasterConfiguration(), "OfflineRegionsSnapshotPool");
+    try {
+      ModifyRegionUtils.editRegions(exec, regions, new ModifyRegionUtils.RegionEditTask() {
+        @Override
+        public void editRegion(final RegionInfo regionInfo) throws IOException {
+          snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), regionInfo);
+        }
+      });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing offline regions of table: " + snapshotTable);
+  }
+
+  private void snapshotMobRegion() throws IOException {
+    RegionInfo mobRegionInfo = MobUtils.getMobRegionInfo(htd.getTableName());
+    snapshotManifest.addMobRegion(mobRegionInfo);
+    status.setStatus("Completed referencing HFiles for the mob region of table: " + snapshotTable);
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    serializer.serialize(SnapshotProcedureStateData
```

Comment:
We do not serialize snapshotManifest here? Then how do we restore it when recovery?